xapian-applications/omega/handler_libarchive.cc

```diff
@@ -0,0 +1,122 @@
+/** @file handler_libarchive.cc
+ * @brief Extract text and metadata using libarchive.
+ */
+/* Copyright (C) 2020 Parth Kapadia
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+#include <config.h>
+#include "handler.h"
+
+#include "metaxmlparse.h"
+#include "opendocparse.h"
+#include "str.h"
+
+#include<archive.h>
+#include<archive_entry.h>
+
+#include <cstring>
+
+using namespace std;
+
+bool
+extract(const string& filename,
+	string& dump,
+	string& title,
+	string& keywords,
+	string& author,
+	string& pages,
+	string& error)
+{
+    try {
+	struct archive* archive_obj;
+	struct archive_entry* entry;
+	int status_code;
+	const char* file = filename.c_str();
+	archive_obj = archive_read_new();
+	archive_read_support_filter_all(archive_obj);
+	archive_read_support_format_all(archive_obj);
+	status_code = archive_read_open_filename(archive_obj, file, 10240);
+
+	if (status_code != ARCHIVE_OK) {
+	    error = "Libarchive failed to open the file " + filename;
+		return false;
+	}
+
+	size_t total;
+	ssize_t size;
+	string s;
+
+	// extracting data from content.xml and styles.xml
+	while (archive_read_next_header(archive_obj, &entry) == ARCHIVE_OK) {
+	    if (strcmp(archive_entry_pathname(entry), "content.xml") == 0) {
+		total = archive_entry_size(entry);
+		char buf1[total];
+		size = archive_read_data(archive_obj, buf1, total);
+
+		if (size <= 0) {
+		    error = "Libarchive was not able to extract data from "
+		    "content.xml";
+		    return false;
+		}
+
+		s += str(buf1);
+		} else if (strcmp(
+			archive_entry_pathname(entry), "styles.xml") == 0) {
+		total = archive_entry_size(entry);
+		char buf2[total];
+		size = archive_read_data(archive_obj, buf2, total);
+
+		if (size <= 0) {
+		    error = "Libarchive was not able to extract data from "
+		    "styles.xml";
+		    return false;
+		}
+
+		s += str(buf2);
+		OpenDocParser parser;
+		parser.parse(s);
+		dump = parser.dump;
```

Comment:
As you cannot assume any order in which files are shown, you cannot parse the content of the file here. It is highly possible that `s` doesn't store the information extracted from `content.xml` at this point so you should parse the information at the end of the program.

Reply:
I have now used OpenDocParser at the end of the file.
@barufa, should I use different strings to store data extracted from content.xml and styles.xml and later append them? 
Or is it alright if they are in a different order?

Reply:
Not sure if it is necessary, but I would try to preserve the order. It would be great to process the file in the same way that we were doing it on `index_file.cc`.

Reply:
It is necessary - that's why `index_file.cc` calls `unzip` separately for `context.xml` and then for `styles.xml`, rather than calling `unzip` once and listing both files - see 5811711bb0f3b9dca7f8be6279290df31fa80397 (which fixes a bug where we were processing them in the order they happened to be inside the zip container).

Reply:
I have changed this, have used different strings for content.xml and styles.xml and appended them to s.